File: day03.py
"""Solution do day 3 of Advent of Code"""

from collections import defaultdict

from get_input import get_input, line_parser
from common import Point


class Wire:
    HEADINGS = {
        "U": Point(0, 1),
        "D": Point(0, -1),
        "R": Point(1, 0),
        "L": Point(-1, 0),
    }

    def __init__(self, path, start=Point(0, 0)):
        self._path = defaultdict(set)
        step = 0
        location = start
        for direction, n_steps in path:
            for _ in range(n_steps):
                step += 1
                location += Wire.HEADINGS[direction]
                self._path[location].add(step)

    def intersections(self, other):
        return set(p for p in self._path if p in other._path)


def part1(wires):
    wire_paths = [Wire(path) for n, path in enumerate(wires)]
    intersections = wire_paths[0].intersections(wire_paths[1])
    return min(abs(point.x) + abs(point.y) for point in intersections)


def part2(wires):
    wire_paths = [Wire(path) for n, path in enumerate(wires)]
    intersections = wire_paths[0].intersections(wire_paths[1])
    return min(
        min(wire_paths[0]._path[i]) + min(wire_paths[1]._path[i])
        for i in intersections
    )


def parse(line):
    return tuple((e[0], int(e[1:])) for e in line.split(','))


TEST1 = """R8,U5,L5,D3\nU7,R6,D4,L4"""
TEST2 = """R75,D30,R83,U83,L12,D49,R71,U7,L72
U62,R66,U55,R34,D71,R55,D58,R83"""

if __name__ == "__main__":
    assert part1(line_parser(TEST1, parse=parse)) == 6
    assert part1(line_parser(TEST2, parse=parse)) == 159
    values = line_parser(get_input(day=3, year=2019), parse=parse)
    print(f"Part 1: {part1(values)}")
    assert part2(line_parser(TEST2, parse=parse)) == 610
    print(f"Part 2: {part2(values)}")

File: day03.rs
use std::collections::{HashMap, HashSet};

use common::get_input;
use common::point::Point;

type WirePath = HashMap<Point, HashSet<usize>>;
type Input = (WirePath, WirePath);

trait Wire {
    fn create_wire(path: &Vec<(String, usize)>) -> Self;
    fn intersections(&self, other: &Self) -> HashSet<Point>;
}

impl Wire for HashMap<Point, HashSet<usize>> {
    fn create_wire(path: &Vec<(String, usize)>) -> Self {
        let mut result = HashMap::new();
        let mut step: usize = 0;
        let mut location: Point = Point::new(0, 0);
        for (direction, steps) in path {
            let diff = match direction.as_ref() {
                "U" => Point::new(0, 1),
                "D" => Point::new(0, -1),
                "R" => Point::new(1, 0),
                "L" => Point::new(-1, 0),
                _ => panic!("Not a known letter"),
            };
            for _ in 0..*steps {
                step += 1;
                location = location + diff;
                let steps = result.entry(location).or_insert(HashSet::new());
                steps.insert(step);
            }
        }
        result
    }

    fn intersections(&self, other: &Self) -> HashSet<Point> {
        let mut intersections: HashSet<Point> = HashSet::new();
        for point in self.keys() {
            if other.contains_key(&point) {
                intersections.insert(point.clone());
            }
        }
        intersections
    }
}

fn part1(paths: &Input) -> usize {
    let intersections = paths.0.intersections(&paths.1);
    let mut result: Option<usize> = None;
    for point in intersections {
        if result.is_none() || result.unwrap() > (point.x.abs() + point.y.abs()) as usize {
            result = Some((point.x.abs() + point.y.abs()) as usize)
        }
    }
    result.unwrap()
}

fn part2(paths: &Input) -> usize {
    let (wire_a, wire_b) = paths;
    let intersections = wire_a.intersections(&wire_b);
    let mut result: Option<usize> = None;
    for point in intersections {
        let local_min = wire_a[&point].iter().min().unwrap() + wire_b[&point].iter().min().unwrap();
        if result.is_none() || local_min < result.unwrap() {
            result = Some(local_min);
        }
    }
    result.unwrap()
}

fn parse(text: &str) -> Input {
    let mut result = Vec::new();
    for line in text.trim().split('\n') {
        let mut row = Vec::new();
        for element in line.split(',') {
            row.push((
                element[0..1].to_string(),
                element[1..].parse().unwrap(),
            ));
        }
        result.push(row);
    }
    assert!(result.len() == 2);
    (HashMap::create_wire(&result[0]), HashMap::create_wire(&result[1]))
}


fn main() {
    let input = parse(&get_input(3, 2019));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
